packages/pds/src/app-view/api/app/bsky/unspecced.ts

```diff
@@ -192,6 +192,69 @@ export default function (server: Server, ctx: AppContext) {
     },
   })
 
+  server.app.bsky.unspecced.putNotificationPushToken({
+    auth: ctx.accessVerifier,
+    handler: async ({ auth, params }) => {
+      const { token, platform, appId, endpoint } = params
+      const {
+        credentials: { did },
+      } = auth
+      const notificationServer = ctx.notifServer
+      try {
+        await notificationServer.registerDeviceForPushNotifications(
+          did,
+          platform,
+          token,
+          endpoint,
+          appId,
+        )
+      } catch (error) {
+        throw new Error('Failed to register device push notification token')
+      }
+    },
+  })
+
+  server.app.bsky.unspecced.registerPushNotificationEndpoint({
+    auth: ctx.accessVerifier,
+    handler: async ({ auth, params, req }) => {
+      const { token, platform, appId, endpoint } = params
+      const {
+        credentials: { did },
+      } = auth
+
+      const { appviewAgent } = ctx
+
+      if (ctx.canProxyWrite()) {
+        try {
+          // TODO: hit the endpoint here so that 3rd party devs can run their own notification servers
+          // const headers = await ctx.serviceAuthHeaders(did)
+          // const called = await appviewAgent.api.xrpc.call(
+          //   'app.bsky.unspecced.putNotificationPushToken',
+          //   {
+          //     token,
+          //     platform,
+          //     appId,
+          //     endpoint,
+          //   },
+          //   headers,
+          // )
+          const res =
+            await appviewAgent.api.app.bsky.unspecced.putNotificationPushToken(
+              {
+                token,
+                platform,
+                appId,
+                endpoint,
+              },
+              await ctx.serviceAuthHeaders(did),
+            )
+        } catch (error) {
+          throw new Error('Failed to register device push notification token')
```

Comment:
Errors like this will get transformed into an opaque internal server error - which won't report anything that went wrong to the client.

Throwing one of the errors from `@atproto/xrpc-server` will do better reporting

Reply:
done